File: lin-compiler/lin_compiler/ScriptRead.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace LIN
{
    static class ScriptRead
    {
        public static bool ReadSource(Script script, string filename)
        {
            // Default script type is textless
            script.Type = ScriptType.Textless;
            Program.PrintLine("[read] reading source file...");

            string[] lines = System.IO.File.ReadAllLines(filename, Encoding.UTF8);
            List<ScriptEntry> ScriptData = [];

            // Regex pattern to match: OpcodeName(args)
            // Handles both quoted strings and regular args
            Regex opcodePattern = new Regex(@"^\s*(\w+|0x[0-9A-Fa-f]+)\s*\((.*)\)\s*$", RegexOptions.Compiled);

            for (int lineNum = 0; lineNum < lines.Length; lineNum++)
            {
                string line = lines[lineNum].Trim();
                if (string.IsNullOrWhiteSpace(line))
                    continue;

                // Skip comment lines
                if (line.StartsWith("#"))
                {
                    continue;
                }

                Match match = opcodePattern.Match(line);
                if (!match.Success)
                {
                    throw new Exception($"[read] error: invalid syntax at line {lineNum + 1}: {line}");
                }

                string opcodeName = match.Groups[1].Value;
                string argsString = match.Groups[2].Value;

                BaseOpcode opcodeDefinition = OpcodeDictionary.GetOpcodeDefinitionByName(opcodeName);
                List<ScriptEntry> entries = opcodeDefinition.ReadSource(argsString, lineNum, script);
                ScriptData.AddRange(entries);
            }

            script.ScriptData = ScriptData;
            return true;
        }

        public static bool ReadCompiled(Script s, byte[] Bytes)
        {
            Program.PrintLine("[read] reading compiled file...");
            s.File = Bytes;
            Program.PrintLine("[read] reading header...");
            s.Type = (ScriptType)BitConverter.ToInt32(s.File, 0x0);
            s.HeaderSize = BitConverter.ToInt32(s.File, 0x4);
            switch (s.Type)
            {
                case ScriptType.Textless:
                    s.FileSize = BitConverter.ToInt32(s.File, 0x8);
                    if (s.FileSize == 0)
                        s.FileSize = s.File.Length;
                    s.TextBlockPos = s.FileSize;
                    s.ScriptData = ReadScriptData(s);
                    break;
                case ScriptType.Text:
                    s.TextBlockPos = BitConverter.ToInt32(s.File, 0x8);
                    s.FileSize = BitConverter.ToInt32(s.File, 0xC);
                    if (s.FileSize == 0)
                        s.FileSize = s.File.Length;
                    s.ScriptData = ReadScriptData(s);
                    s.TextEntries = BitConverter.ToInt32(s.File, s.TextBlockPos);
                    ReadTextEntries(s);
                    break;
                default:
                    throw new Exception("[read] error: unknown script type.");
            }

            return true;
        }

        private static List<ScriptEntry> ReadScriptData(Script s)
        {
            Program.PrintLine("[read] reading script data...");
            List<ScriptEntry> ScriptData = [];
            for (int i = s.HeaderSize; i < s.TextBlockPos; i++)
            {
                if (s.File[i] == 0x70)
                {
                    i++;
                    ScriptEntry e = new()
                    {
                        Opcode = s.File[i]
                    };

                    int ArgCount = OpcodeDictionary.GetOpcodeArgCount(e.Opcode);
                    if (ArgCount == -1)
                    {
                        // Vararg
                        List<byte> Args = [];
                        while (s.File[i + 1] != 0x70)
                        {
                            Args.Add(s.File[i + 1]);
                            i++;
                        }
                        e.Args = [.. Args];
                        ScriptData.Add(e);
                        continue;
                    }
                    else
                    {
                        e.Args = new byte[ArgCount];
                        for (int a = 0; a < e.Args.Length; a++)
                        {
                            e.Args[a] = s.File[i + 1];
                            i++;
                        }
                        ScriptData.Add(e);
                    }
                }
                else
                {
                    // EOF?
                    while (i < s.TextBlockPos)
                    {
                        if (s.File[i] != 0x00)
                        {
                            throw new Exception("[read] error: expected 0x70, got 0x" + s.File[i].ToString("X2") + ".");
                        }
                        i++;
                    }
                    return ScriptData;
                }
            }
            return ScriptData;
        }

        private static void ReadTextEntries(Script s)
        {
            Program.PrintLine("[read] reading text entries...");
            List<int> TextIDs = new(s.TextEntries);
            for (int i = 0; i < s.ScriptData.Count; i++)
            {
                if (s.ScriptData[i].Opcode == 0x02)
                {
                    byte first = s.ScriptData[i].Args[0];
                    byte second = s.ScriptData[i].Args[1];
                    // Big-endian: MSB first, LSB second (as stored in file)
                    int TextID = (first << 8) | second;

                    if (TextID >= s.TextEntries)
                    {
                        throw new Exception("[read] error: text id out of range.");
                    }

                    TextIDs.Add(TextID);
                    int TextPos = BitConverter.ToInt32(s.File, s.TextBlockPos + (TextID + 1) * 4);
                    int NextTextPos = BitConverter.ToInt32(s.File, s.TextBlockPos + (TextID + 2) * 4);
                    if (TextID == s.TextEntries - 1) NextTextPos = s.FileSize - s.TextBlockPos;

                    // Read the string from the byte array
                    string text = Encoding.Unicode.GetString(s.File, s.TextBlockPos + TextPos, NextTextPos - TextPos);

                    // Remove the BOM if present
                    if (text.StartsWith("\uFFFE"))
                    {
                        // Remove the BOM
                        text = text.Substring(1); // Skip the first character
                    }

                    // Assign the cleaned text back to the ScriptData
                    s.ScriptData[i].Text = text;
                }
                else
                {
                    s.ScriptData[i].Text = null;
                }
            }
        }

    }
}

File: projects/lin-compiler/src/scriptRead.ts
import { readFile } from "node:fs/promises";
import { BinaryError, SourceError } from "./errors.ts";
import { OP_TEXT, OPCODE_MARKER } from "./opcodes/ids.ts";
import { getOpcode, getOpcodeByName, hexOpcodeName, parseHexOpcodeName } from "./opcodes/opcodeDictionary.ts";
import { ParamType, parseArg, splitArgs } from "./parameter.ts";
import { type Script, type ScriptEntry, ScriptType } from "./script.ts";

// ---------------------------------------------------------------------------
// .linscript source
// ---------------------------------------------------------------------------

/** Matches `OpcodeName(args)` or `0xNN(args)`, capturing the name and the raw argument text. */
const OPCODE_LINE = /^(\w+)\s*\((.*)\)$/;

/** Parse `.linscript` source text. Blank lines and `#` comments are ignored. */
export function readSource(source: string): Script {
  const entries: ScriptEntry[] = [];

  splitLines(source).forEach((rawLine, index) => {
    const line = index + 1;
    const text = rawLine.trim();
    if (text.length === 0 || text.startsWith("#")) {
      return;
    }

    const match = OPCODE_LINE.exec(text);
    if (match === null) {
      throw new SourceError(line, `invalid syntax: ${text}`);
    }
    const [, name, argsText] = match;
    entries.push(...parseOpcodeLine(name, argsText, line));
  });

  return { entries };
}

export async function readSourceFile(path: string): Promise<Script> {
  return readSource(await readFile(path, "utf8"));
}

function parseOpcodeLine(name: string, argsText: string, line: number): ScriptEntry[] {
  const opcode = getOpcodeByName(name);
  if (opcode !== undefined) {
    return opcode.parseSource(argsText, line);
  }

  // An unregistered `0xNN(a, b, c)` is written by the decompiler for unknown opcodes; take its bytes verbatim
  const rawId = parseHexOpcodeName(name);
  if (rawId !== undefined) {
    const args = splitArgs(argsText).flatMap((value) => parseArg(ParamType.Byte, value, line));
    return [{ opcode: rawId, args }];
  }

  throw new SourceError(line, `unknown opcode '${name}'`);
}

function splitLines(source: string): string[] {
  const text = source.startsWith("\uFEFF") ? source.slice(1) : source;
  return text.split(/\r\n|\r|\n/);
}

// ---------------------------------------------------------------------------
// compiled .lin bytes
// ---------------------------------------------------------------------------

/**
 * Parse a compiled `.lin` file.
 *
 * Layout (all integers little-endian int32):
 *
 *     Textless: type=1, headerSize, fileSize, script data...
 *     Text:     type=2, headerSize, textBlockPos, fileSize, script data..., text block
 *
 * Script data is a sequence of `0x70 <opcode> <args...>` records, zero-padded to the text block.
 * The text block is a count, `count + 1` offsets relative to the block start, then UTF-16LE strings.
 */
export function readCompiled(bytes: Uint8Array): Script {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const int32 = (offset: number) => view.getInt32(offset, true);

  const type = int32(0x0);
  const headerSize = int32(0x4);

  switch (type) {
    case ScriptType.Textless: {
      const fileSize = int32(0x8) || bytes.length;
      return { entries: readScriptData(bytes, headerSize, fileSize) };
    }
    case ScriptType.Text: {
      const textBlockPos = int32(0x8);
      const fileSize = int32(0xc) || bytes.length;
      const entries = readScriptData(bytes, headerSize, textBlockPos);
      attachTextEntries(entries, bytes, textBlockPos, fileSize);
      return { entries };
    }
    default:
      throw new BinaryError(`unknown script type ${type}`);
  }
}

export async function readCompiledFile(path: string): Promise<Script> {
  return readCompiled(new Uint8Array(await readFile(path)));
}

function readScriptData(bytes: Uint8Array, start: number, end: number): ScriptEntry[] {
  const entries: ScriptEntry[] = [];
  let pos = start;

  while (pos < end) {
    if (bytes[pos] !== OPCODE_MARKER) {
      expectZeroPadding(bytes, pos, end);
      break;
    }
    const id = bytes[pos + 1];
    pos += 2;

    const opcode = getOpcode(id);
    const argEnd = opcode === undefined || opcode.variadic ? findNextMarker(bytes, pos, end) : pos + opcode.argByteCount;

    entries.push({ opcode: id, args: Array.from(bytes.subarray(pos, argEnd)) });
    pos = argEnd;
  }

  return entries;
}

/** Once the opcode records stop, only zero bytes may remain before `end`. */
function expectZeroPadding(bytes: Uint8Array, from: number, end: number): void {
  for (let pos = from; pos < end; pos++) {
    if (bytes[pos] !== 0x00) {
      throw new BinaryError(`expected opcode marker ${hexOpcodeName(OPCODE_MARKER)} at offset ${pos}, got ${hexOpcodeName(bytes[pos])}`);
    }
  }
}

function findNextMarker(bytes: Uint8Array, from: number, end: number): number {
  let pos = from;
  while (pos < end && bytes[pos] !== OPCODE_MARKER) {
    pos++;
  }
  return pos;
}

function attachTextEntries(entries: ScriptEntry[], bytes: Uint8Array, textBlockPos: number, fileSize: number): void {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const textCount = view.getInt32(textBlockPos, true);
  const offsetAt = (textId: number) => view.getInt32(textBlockPos + 4 + textId * 4, true);
  const buffer = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);

  for (const entry of entries) {
    if (entry.opcode !== OP_TEXT) {
      continue;
    }
    const textId = (entry.args[0] << 8) | entry.args[1];
    if (textId >= textCount) {
      throw new BinaryError(`text id ${textId} out of range (${textCount} entries)`);
    }

    const start = textBlockPos + offsetAt(textId);
    const end = textId === textCount - 1 ? fileSize : textBlockPos + offsetAt(textId + 1);
    const text = buffer.toString("utf16le", start, end);
    // Drop a byte-reversed BOM if one leads the entry
    entry.text = text.startsWith("\uFFFE") ? text.slice(1) : text;
  }
}
